lib/jekyll/commands/serve.rb

```diff
@@ -33,26 +48,117 @@ def init_with_program(prog)
             cmd.action do |_, opts|
               opts["serving"] = true
               opts["watch"  ] = true unless opts.key?("watch")
-              config = opts["config"]
-              Build.process(opts)
-              opts["config"] = config
-              Serve.process(opts)
+              opts["reload_port"] = LIVERELOAD_PORT unless opts.key?("reload_port")
+
+              validate_options(opts)
+              start(opts)
             end
           end
         end
 
         #
 
+        def start(opts)
+          config = opts["config"]
+          @reload_reactor = nil
+          register_reload_hooks(opts) if opts["livereload"]
+          Build.process(opts)
+          opts["config"] = config
+          Serve.process(opts)
+        end
+
+        #
+
+        # rubocop:disable Metrics/AbcSize
         def process(opts)
           opts = configuration_from_options(opts)
           destination = opts["destination"]
           setup(destination)
 
-          server = WEBrick::HTTPServer.new(webrick_opts(opts)).tap { |o| o.unmount("") }
-          server.mount(opts["baseurl"], Servlet, destination, file_handler_opts)
-          Jekyll.logger.info "Server address:", server_address(server, opts)
-          launch_browser server, opts if opts["open_url"]
-          boot_or_detach server, opts
+          # Need some way of communicating between the stop and start callbacks
+          @running = Queue.new
+
+          if opts["livereload"]
+            @reload_reactor.start(opts)
+          end
+
+          @server = WEBrick::HTTPServer.new(webrick_opts(opts)).tap { |o| o.unmount("") }
+
+          @server.mount("#{opts["baseurl"]}/__livereload",
+            WEBrick::HTTPServlet::FileHandler, LIVERELOAD_DIR)
+          @server.mount(opts["baseurl"], Servlet, destination, file_handler_opts)
+
+          Jekyll.logger.info "Server address:", server_address(@server, opts)
+          launch_browser @server, opts if opts["open_url"]
+          boot_or_detach @server, opts
+        end
+
+        #
+
+        def running?
+          !(@running.nil? || @running.empty?)
+        end
+
+        #
+
+        def shutdown
+          @server.shutdown if running?
+        end
+
+        private
+        # rubocop:disable Metrics/CyclomaticComplexity
```

Comment:
> This is a no-go. Please cleanup your method, it's pretty complex.

Okay, the complexity decreased below the threshold with the latest commit, but it is still fairly ugly.  I'm not sure about the best strategy.  Maybe just drop the warning that `--livereload` and `--no-watch` don't make any sense?